core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -642,6 +643,24 @@ private[log] class Cleaner(val id: Int,
   }
 
   /**
+   * Estimate what the last offset is for the current segment because we can't be sure
+   * that the index is returning a valid value in all cases.
```

Comment:
How about changing the comment to the following.

We want to get the last offset in the first log segment in segs.
LogSegment.nextOffset() gives the exact last offset in a segment, but can be expensive since it requires scanning the segment from the last index entry. Therefore, we estimate the last offset of the first log segment by using the base offset of the next segment in the list. If the next segment doesn't exist, firstUncleanableOffset will be used.